services/registry/test/models/user.spec.js

```diff
@@ -0,0 +1,178 @@
+  return await User.objects.create({
+    name,
+    email
+  });
+}
+
+const USER_ERRORS = {
+  NO_NAME: 'child "name" fails because ["name" is not allowed to be empty]'
+};
+
+describe('User', () => {
+  it(
+    'can create a user with name and email',
+
+    providePostgres(async () => {
+      const user = await addNewUser('foo bar', 'bar@foo.com');
+      demand(user.name).to.be('foo bar');
+      demand(user.email).to.be('bar@foo.com');
+      demand(user.active).to.be.truthy();
+    })
+  );
+
+  it(
+    'throws an error when a user exists',
+    providePostgres(async () => {
+      let error = undefined;
+
+      try {
+        await addNewUser('foo bar', 'bar@foo.com');
+        await addNewUser('foo bar', 'bar@foo.com');
+      } catch (e) {
+        error = e.message;
+      }
+
+      demand(error).to.be('Usernames must be unique.');
+    })
+  );
+
+  it(
+    'throws an error when usernames are not at least 1 character',
+    providePostgres(async () => {
+      let error = undefined;
+
+      try {
+        await addNewUser('', 'bar@foo.com');
+      } catch (e) {
+        error = e.message;
+      }
+
+      demand(error).to.be(USER_ERRORS.NO_NAME);
+    })
+  );
+
+  describe('signup', () => {
+    let prevOAuthEnvVar;
+    before(() => {
+      // Capture for reset later when env is changed
+      prevOAuthEnvVar = process.env.OAUTH_PASSWORD;
+    });
+
+    after(() => {
+      // reset
+      process.env.OAUTH_PASSWORD = prevOAuthEnvVar;
+    });
+
+    it(
+      'can sign a user up with valid parameters',
+      providePostgres(async () => {
+        const newSignup = await User.signup('foo', 'bar@foo.com', false);
+
+        demand(newSignup instanceof User).to.be.truthy();
+        demand(newSignup.active).to.be.truthy();
+      })
+    );
+
+    it(
+      'throws an error given an invalid name',
+      providePostgres(async () => {
+        let error = undefined;
+        try {
+          await User.signup('', 'bar@foo.com', false);
+        } catch (e) {
+          error = e.message;
+        }
+
+        demand(error).to.be(USER_ERRORS.NO_NAME);
+      })
+    );
+
+    it(
+      'creates a namespace for the user',
+      providePostgres(async () => {
+        const name = 'foo bar';
+        const newUser = await User.signup(name, 'bar@foo.com', false);
+
+        const userNamespace = await Namespace.objects.get({
+          active: true,
+          name: newUser.name
+        });
+
+        demand(userNamespace instanceof Namespace).to.be.truthy();
+        demand(userNamespace.active).to.be.truthy();
+      })
+    );
+
+    it(
+      'creates a namespace with a host for the user',
+      providePostgres(async () => {
+        const name = 'foo bar';
+        await User.signup(name, 'bar@foo.com', false);
+
+        // Fetch the user's namespace that was created during signup.
+        const userNamespace = await Namespace.objects.filter({
+          'host.id': 1, // TODO: don't hardcode?
```

Comment:
I'd be tempted to add a helper function to `Host` to grab the current hostname so we can get condense all of the `process.env.EXTERNAL_HOST.replace(/https?:\/\//, '')` bits into a single `Host.current()` call. That would let us write `'host.name': Host.current()` here, for example!

(This can definitely be left for a separate PR.)

Reply:
Oh, that would be nice! When I first looked at this, I had to do some logging to figure out what 1 actually meant. So a helper would be ... helpful :slightly_smiling_face: